### PR TITLE
Caching and related restructuring

### PR DESCRIPTION
**Pull request by cobusc**

### CODE REVIEW

File: httpd.py [hunk 1]
```diff
@@ -1,29 +1,46 @@
...
+    return [TECH_ADMIN]
+
+
 if __name__ == "__main__":
-    if settings.INSECURE:
+    if settings.INSECURE:  # TODO: Remove before going to prod
         print("*" * 29)
         print("* Running in insecure mode! *")
         print("*" * 29)
+        setattr(utils, "get_user_roles_for_site", mocked_return)
```
Comment (AltusBarry) [inline httpd.py]: I assume this also needs the TODO comment?

Comment (AltusBarry) [reply httpd.py]: :woman_facepalming: Yeah nevermind. I swear I am not always this slow.

File: management_layer/integration.py [hunk 2]
```diff
@@ -1165,6 +1187,7 @@ class Implementation(AbstractStubClass):
         }
 
     @staticmethod
+    @require_permissions(all, [("urn:ge:identity_provider:user", "delete")])
```
Comment (AltusBarry) [inline management_layer/integration.py]: We might need an extra ticket to ensure all data related to the user is cleared on auth service in the event of user deletion.

Comment (cobusc) [reply management_layer/integration.py]: Correct. We may have to do a special `delete_all_core_user_data` API call that will remove entries from all 3 components as well as send out emails, etc.

File: management_layer/permission/utils.py [hunk 3]
```diff
@@ -124,28 +94,47 @@ def roles_have_permissions(
...
-    return operator(
-        # Any role can provide the permission.
-        any(role_has_permission(role, permission, resource, nocache)
-            for role in roles)
-        for resource, permission in resource_permissions
-    )
-
-
-def user_has_permissions(
+    #return operator(
```
Comment (AltusBarry) [inline management_layer/permission/utils.py]: I suppose this commented code can be removed?

File: management_layer/permission/utils.py [hunk 4]
```diff
@@ -212,25 +204,36 @@ def get_role_resource_permissions(
     resource_id = resource if type(resource) is int else \
         RESOURCE_URN_TO_ID_MAP[resource]
 
-    key = "{}:{}:{}".format(MKP_ROLE_RESOURCE_PERMISSION, role_id, resource_id)
-    role_resource_permissions = None if nocache else MEMCACHE.get(key)
+    key = bytes("{}:{}:{}".format(MKP_ROLE_RESOURCE_PERMISSION, role_id, resource_id),
```
Comment (AltusBarry) [inline management_layer/permission/utils.py]: Would using the `f-string` notation make this easier to maintain?

File: management_layer/permission/utils.py [hunk 5]
```diff
@@ -255,62 +258,73 @@ def get_all_user_roles(
...
     used in exceptional circumstances.
+    :param request: The request associated with this call
     :param user: The user for which to get roles
     :param nocache: Bypass the cache if True
     :return: A dictionary containing the roles assigned to the specified user on
        the domains and sites in the system.
     """
-    key = "{}:{}".format(MKP_USER_ROLES, user.hex)
-    user_roles = None if nocache else MEMCACHE.get(key)
+    key = bytes("{}:{}".format(MKP_USER_ROLES, user.hex), encoding="utf8")
```
Comment (AltusBarry) [inline management_layer/permission/utils.py]: `f-string` notation might save a few characters here as well.

File: requirements.txt [hunk 6]
```diff
@@ -1,9 +1,9 @@
 -e git+https://github.com/praekelt/swagger-django-generator.git@master#egg=swagger-django-generator
 -e git+https://github.com/praekelt/apitools@master#egg=apitools
 python_dateutil == 2.6.0
-pymemcache < 2.0.0 # To be replaced by aiomcache
+aiomcache
```
Comment (AltusBarry) [inline requirements.txt]: Should we perhaps pin the version for this?

**Comment by AltusBarry** [comment]
+1 Only some minor comments

File: requirements.txt (hunk 6 above)
Comment (cobusc) [reply requirements.txt]: Yes, later.

**Pull request merged**